src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs

```diff
@@ -2299,35 +2300,164 @@ internal static void VarDecMod(ref DecCalc d1, ref DecCalc d2)
                     }
+                } while (scale < 0);
+            }
+
+            private static unsafe void VarDecModFull(ref DecCalc d1, ref DecCalc d2, int scale)
+            {
+                // Divisor has bits set in the upper 64 bits.
+                //
+                // Divisor must be fully normalized (shifted so bit 31 of the most significant ULONG is 1). 
+                // Locate the MSB so we know how much to normalize by. 
+                // The dividend will be shifted by the same amount so the quotient is not changed.
+                //
+                uint tmp = d2.High;
+                if (tmp == 0)
+                    tmp = d2.Mid;
+                int shift = X86.Lzcnt.IsSupported ? (int)X86.Lzcnt.LeadingZeroCount(tmp) : LeadingZeroCount(tmp);
+
+                Buf28 b;
+                _ = &b; // workaround for CS0165
+                b.Buf24.Low64 = d1.Low64 << shift;
+                b.Buf24.Mid64 = (d1.Mid + ((ulong)d1.High << 32)) >> (32 - shift);
+
+                // The dividend might need to be scaled up to 221 significant bits.
+                uint high = 3;
+                while (scale < 0)
+                {
+                    uint power = scale <= -MaxInt32Scale ? TenToPowerNine : s_powers10[-scale];
+                    uint* buf = (uint*)&b;
+                    ulong tmp64 = UInt32x32To64(b.Buf24.U0, power);
+                    b.Buf24.U0 = (uint)tmp64;
+                    for (int i = 1; i <= high; i++)
+                    {
+                        tmp64 >>= 32;
+                        tmp64 += UInt32x32To64(buf[i], power);
+                        buf[i] = (uint)tmp64;
+                    }
+                    if (tmp64 > int.MaxValue)
+                    {
+                        // The high bit of the dividend must not be set.
+                        Debug.Assert(high + 1 < b.Length);
+                        buf[++high] = (uint)(tmp64 >> 32);
+                    }
+
+                    scale += MaxInt32Scale;
+                }
+
+                if (d2.High == 0)
+                {
+                    ulong divisor = d2.Low64 << shift;
+                    switch (high)
+                    {
+                        case 6:
+                            Div96By64(ref *(Buf12*)&b.Buf24.U4, divisor);
+                            goto case 5;
+                        case 5:
+                            Div96By64(ref *(Buf12*)&b.Buf24.U3, divisor);
+                            goto case 4;
+                        case 4:
+                            Div96By64(ref *(Buf12*)&b.Buf24.U2, divisor);
+                            break;
+                    }
+                    Div96By64(ref *(Buf12*)&b.Buf24.U1, divisor);
+                    Div96By64(ref *(Buf12*)&b, divisor);
+
+                    d1.Low64 = b.Buf24.Low64 >> shift;
+                    d1.High = 0;
+                }
+                else
+                {
+                    Buf12 bufDivisor;
```

Comment:
would it make sense to just add constructor which takes Low64 and U2?

Reply:
It wouldn't help here because it's still going to complain that all fields must be fully assigned (in the constructor).

Reply:
is adding `= default(Buf12)` slowing it down? (by how much)

Reply:
Using `= default` instead of `_ = &variable` adds about 2% overhead.